### PR TITLE
Refactored matplotlib axis setting for easier reusage

### PR DESCRIPTION
**Pull request by philippjfr**
Not ready yet but this PR begins to refactor some of the code in setting up the axes limits, labels, ticks, aspect etc. to make reusing this code easier in custom extensions with different types of axes.

### CODE REVIEW

**Comment by jlstevens** [comment]
All these changes look good.

@philippjfr  That said, can we agree to have a freeze on the plotting code once this is merged?

I really feel we need a chance to document as much of the API (including internal API) that you consider stable. This will take the pressure off you and allow other people to start contributing by writing their own plotting classes. I don't think you should be the sole maintainer of the plotting module forever!

In other words, I think all these refactors have pushed your current [bus factor](https://en.wikipedia.org/wiki/Bus_factor) a bit too high. :-)

**Comment by jlstevens** [comment]
I would also like to link to PR #438. That seems the right place to start thinking about the API in order to finally document it properly.

**Comment by philippjfr** [comment]
Yes, there will be a freeze on the plotting code until we have sufficiently documented the current implementation. Both the matplotlib and bokeh element plots should now be pretty stable and after documenting both we can consider how to improve other aspects of the plotting code including improved layouts, unifying some of the options across backends and then finally document how to create an entire backend.

**Comment by philippjfr** [comment]
The current test failures are as follows and are all expected:

* Columnar Data:
  - 15/16/22: Axis label changes
* Composing_Data:
  - 0/3: GridSpace gridlines reenabled
* Containers:
  - 9: GridSpace gridlines reenabled
* Elements:
  - 8/10: Axis label changes
  - 11/21: Tiny axis frame change

The axis label change is to provide more consistent handling for axes representing more than one dimension (e.g. on Bars and ErrorBars types). The GridSpace line change reverts an earlier change where I'd accidentally removed gridlines completely when the axis itself is hidden. The axis frame change is down to adjoined plots, where the edge where the adjoined plot attaches was automatically added even if not specified. I'll rebuild the data soon and then this PR is ready to review and merge.

**Comment by philippjfr** [comment]
Tests passing, ready to review.

File: holoviews/plotting/mpl/element.py [hunk 1]
```diff
@@ -105,7 +107,8 @@ class ElementPlot(GenericElementPlot, MPLPlot):
     # Element Plots should declare the valid style options for matplotlib call
     style_opts = []
 
-    _suppressed = [Table, NdElement, Collator, Columns, ItemTable]
+    # Whether plot has axes, disables setting axis limits, labels and ticks
+    _axes = True
```
Comment (jlstevens) [inline holoviews/plotting/mpl/element.py]: I would consider renaming this attribute ``_has_axes`` to indicate it is a boolean predicate.

File: holoviews/plotting/util.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
+from __future__ import unicode_literals
```
Comment (jlstevens) [inline holoviews/plotting/util.py]: Not sure how this import relates to the PR but seems like a good idea anyway...

File: holoviews/plotting/mpl/element.py (hunk 1 above)
Comment (philippjfr) [inline holoviews/plotting/mpl/element.py]: Good plan, I considered adding a separate baseclass, but didn't want to change too much.

File: holoviews/plotting/util.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
+
```
Comment (philippjfr) [inline holoviews/plotting/util.py]: The separator declared in the function below has to be unicode.

**Comment by jlstevens** [comment]
I've had a look through and I am happy to merge if you are happy with it. My only other suggestion is to run flakes on your branch because there is quite a bit changed and flakes might help you catch bugs.

**Comment by philippjfr** [comment]
Good plan, I'll apply your ``has_axes`` suggestion and run a few linters over it.

**Comment by philippjfr** [comment]
Made that fix and ran pyflakes, apart from superfluous imports there were a few issues that were introduced in a previous PR I think.

**Comment by jlstevens** [comment]
Ok..the tests are passing. Merging!
